runtime/lua/vim/lsp/codelens.lua

```diff
@@ -0,0 +1,223 @@
+local util = require('vim.lsp.util')
+local api = vim.api
+local M = {}
+
+--- bufnr -> client_id -> lenses
+local lens_cache_by_buf = setmetatable({}, {
+  __index = function(t, b)
+    local key = b > 0 and b or api.nvim_get_current_buf()
+    return rawget(t, key)
+  end
+})
+
+local namespaces = setmetatable({}, {
+  __index = function(t, key)
+    local value = api.nvim_create_namespace('vim_lsp_codelens:' .. key)
+    rawset(t, key, value)
+    return value
+  end;
+})
+
+--@private
+M.__namespaces = namespaces
+
+
+--- Return all lenses for the given buffer
+---@return table (`CodeLens[]`)
+function M.get(bufnr)
+  local lenses_by_client = lens_cache_by_buf[bufnr]
+  if not lenses_by_client then return {} end
+  local lenses = {}
+  for _, client_lenses in pairs(lenses_by_client) do
+    vim.list_extend(lenses, client_lenses)
+  end
+  return lenses
+end
+
+
+--@private
+local function execute_lens(lens, bufnr, client_id)
+  local line = lens.range.start.line
+  api.nvim_buf_clear_namespace(bufnr, namespaces[client_id], line, line + 1)
+
+  -- Need to use the client that returned the lens → must not use buf_request
+  local client = vim.lsp.get_client_by_id(client_id)
+  assert(client, 'Client is required to execute lens, client_id=' .. client_id)
+  client.request('workspace/executeCommand', lens.command, function(...)
+    local result = vim.lsp.handlers['workspace/executeCommand'](...)
+    M.refresh()
+    return result
+  end, bufnr)
+end
+
+
+--- Run the code lens in the current line
+---
+function M.run()
```

Comment:
Is there any reason we might want to allow this to take a buffer/window handle?

Reply:
Can't think of a good reason. If you wanted to call this from somewhere else you'd also need to provide the line number. Not sure in which situation you'd know which lens to run from elsewhere?

If a use-case does pop up it is easier to extend later than to remove in case we anticipate the wrong thing.